src/tools/rbd_mirror/ImageMap.cc

```diff
@@ -0,0 +1,663 @@
+        m_image_info.erase(global_image_id);
+        m_image_action.release_state(global_image_id);
+      };
+
+      delete on_unmap;
+      delete on_purge;
+    });
+
+  ImageAction::Action action = ImageAction::Action::create_remove_action(
+    on_unmap, on_purge, on_finish);
+  append_and_schedule_action(global_image_id, action);
+}
+
+template <typename I>
+void ImageMap<I>::schedule_shuffle_action(const std::string &global_image_id,
+                                          const std::string &source_instance_id,
+                                          const std::string &target_instance_id) {
+  assert(m_lock.is_locked());
+
+  dout(20) << ": global_image_id=" << global_image_id << ", source="
+           << source_instance_id << ", target=" << target_instance_id
+           << dendl;
+
+  Context *on_unmap = new SubActionUnmapContext(this, global_image_id, source_instance_id);
+  Context *on_map = new SubActionMapContext(this, global_image_id, target_instance_id);
+
+  Context *on_finish = new FunctionContext(
+    [this, global_image_id, source_instance_id,
+     target_instance_id, on_map, on_unmap](int r) {
+      assert(m_lock.is_locked());
+
+      if (r < 0) {
+        derr << ": failed to shuffle global_image_id=" << global_image_id
+             << " from instance_id=" << source_instance_id << " to instance_id="
+             << target_instance_id << dendl;
+      }
+
+      set_retry_instance_id(global_image_id, Policy<I>::UNMAPPED_INSTANCE_ID);
+
+      delete on_map;
+      delete on_unmap;
+    });
+
+  ImageAction::Action action = ImageAction::Action::create_shuffle_action(
+    on_map, on_unmap, on_finish);
+  append_and_schedule_action(global_image_id, action);
+}
+
+template <typename I>
+void ImageMap<I>::update_images_added(const std::string &mirror_uuid,
+                                      const std::set<std::string> &global_image_ids) {
+  dout(20) << dendl;
+  assert(m_lock.is_locked());
+
+  for (auto const &global_image_id : global_image_ids) {
+    bool schedule_update = false;
+
+    std::string instance_id = m_policy->lookup(global_image_id);
+    if (instance_id == Policy<I>::UNMAPPED_INSTANCE_ID) {
+      instance_id = m_policy->map(global_image_id);
+
+      schedule_update = true;
+      m_image_action.init_state(global_image_id);
+      m_image_info.emplace(global_image_id, Info());
+    }
+
+    add_peer(global_image_id, mirror_uuid);
+    if (schedule_update) {
+      schedule_add_action(global_image_id, instance_id);
+    }
+  }
+}
+
+template <typename I>
+void ImageMap<I>::update_images_removed(const std::string &mirror_uuid,
+                                        const std::set<std::string> &global_image_ids) {
+  dout(20) << dendl;
+  assert(m_lock.is_locked());
+
+  Updates to_remove;
+
+  for (auto const &global_image_id : global_image_ids) {
+    std::string instance_id = m_policy->lookup(global_image_id);
+    assert(instance_id != Policy<I>::UNMAPPED_INSTANCE_ID);
```

Comment:
Nit: probably should just continue if the instance is already unmapped. It might be possible for a duplicate remove event to fire.